File: ch9/SparseTable.java
import java.util.*;

class RMQ {                                          // Range Minimum Query
  private int[] _A = new int[1000];          // adjust this value as needed
  private int[][] SpT = new int[1000][10];

  public RMQ(int n, int[] A) { // constructor as well as pre-processing routine
    for (int i = 0; i < n; i++) {
      _A[i] = A[i];
      SpT[i][0] = i; // RMQ of sub array starting at index i + length 2^0=1
    }
    // the two nested loops below have overall time complexity = O(n log n)
    for (int j = 1; (1<<j) <= n; j++) // for each j s.t. 2^j <= n, O(log n)
      for (int i = 0; i + (1<<j) - 1 < n; i++)    // for each valid i, O(n)
        if (_A[SpT[i][j-1]] < _A[SpT[i+(1<<(j-1))][j-1]])            // RMQ
          SpT[i][j] = SpT[i][j-1];    // start at index i of length 2^(j-1)
        else                  // start at index i+2^(j-1) of length 2^(j-1)
          SpT[i][j] = SpT[i+(1<<(j-1))][j-1];
  }

  public int query(int i, int j) {
    int k = (int)Math.floor(Math.log((double)j-i+1) / Math.log(2.0)); // 2^k <= (j-i+1)
    if (_A[SpT[i][k]] <= _A[SpT[j-(1<<k)+1][k]]) return SpT[i][k];
    else                                         return SpT[j-(1<<k)+1][k];
} };

class SparseTable {
  public static void main(String[] args) {
    // same example as in chapter 2: segment tree
    int n = 7;
    int[] A = new int[] {18, 17, 13, 19, 15, 11, 20};
    RMQ rmq = new RMQ(n, A);
    for (int i = 0; i < n; i++)
      for (int j = i; j < n; j++)
        System.out.printf("RMQ(%d, %d) = %d\n", i, j, rmq.query(i, j));
  }
}

File: ch9/SparseTable.ml
open Printf

module RMQ = struct                                         (* Range Minimum Query *)
  type t =
    { l2: int array;
      a: int array;
      spt: int array array;
    }

  let init n a =
    let l2 = Array.make (n + 1) 0 in
    begin
      try
        for i = 0 to max_int do
          for j = 1 lsl i to 1 lsl (i + 1) - 1 do
            if j > n then raise Exit;
            l2.(j) <- i
          done
        done
      with Exit -> ()
    end;
    let l2_n = l2.(n) in
    (* initialization *)
    let spt = Array.make_matrix (l2_n + 1) n 0 in
    spt.(0) <- Array.init n (fun i -> i);               (* RMQ of sub array [j..j] *)

    (* the two nested loops below have overall time complexity = O(n log n) *)
    for i = 1 to l2_n do                     (* for each i s.t. 2^i <= n, O(log n) *)
      for j = 0 to n - 1 lsl i do                        (* for each valid j, O(n) *)
        let x = spt.(i-1).(j) in                            (* covers [j..j+2^i-1] *)
        let y = spt.(i-1).(j + 1 lsl (i-1)) in  (* covers [j+(1<<(i-1))..j+(1<<i)] *)
        spt.(i).(j) <- if a.(x) <= a.(y) then x else y
      done
    done;
    {l2; a; spt}

  let query i j {l2; a; spt} =
    let k = l2.(j - i + 1) in                                    (* 2^k <= (j-i+1) *)
    let x = spt.(k).(i) in                                  (* covers [i..i+2^k-1] *)
    let y = spt.(k).(j - 1 lsl k + 1) in                    (* covers [j-2^k+1..j] *)
    if a.(x) <= a.(y) then x else y
end

let () =
  (* same example as in chapter 2: segment tree *)
  let a = [|18; 17; 13; 19; 15; 11; 20|] in
  let n = Array.length a in
  let rmq = RMQ.init n a in
  for i = 0 to n - 1 do
    for j = i to n - 1 do
      printf "RMQ(%d, %d) = %d\n" i j (RMQ.query i j rmq)
    done
  done
